### ISSUE TITLE
gcc-9 printf format warnings in lib557.c

### ISSUE THREAD

**Issue by bagder** [curl/curl#15008]
### I did this

Checked the CI builds for current master. This is a [failed build](https://app.circleci.com/pipelines/github/curl/curl/10241/workflows/50290dff-642d-4ed0-a4ce-60a20c92f03b/jobs/107790?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link&utm_content=summary).

It seems to be using gcc-9. 

This is a false positive because our private implementation does not have a problem with this construct.

~~~
lib557.c: In function ‘test_float_formatting’:
lib557.c:1408:37: error: ‘%*f’ directive output of 2147483648 bytes exceeds ‘INT_MAX’ [-Werror=format-overflow=]
 1408 |   curl_msnprintf(buf, sizeof(buf), "%*f", INT_MIN, 9.1);
      |                                     ^~~
lib557.c:1408:3: note: ‘curl_msnprintf’ output 2147483649 bytes
 1408 |   curl_msnprintf(buf, sizeof(buf), "%*f", INT_MIN, 9.1);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
~~~

### I expected the following

No build errors

### curl/libcurl version

current git master 7b75bd52f6187

### operating system

Linux CI

**Comment by vszakats** [comment]
It failed here first:
https://app.circleci.com/pipelines/github/curl/curl/10226/workflows/87642ee9-cda6-4916-8206-c82aac5f595e
after: https://github.com/curl/curl/commit/71cf0d1fca9e1f53524e1545ef0c08d174458d80 #14772

Question why the bundle test build makes this warnings enabled, because it's not enabled by any compiler option:
[godbolt link](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:13,endLineNumber:42,positionColumn:13,positionLineNumber:42,selectionStartColumn:2,selectionStartLineNumber:42,startColumn:2,startLineNumber:42),source:'%23include+%3Cstdarg.h%3E%0A%23include+%3Cstdio.h%3E%0A%23include+%3Climits.h%3E%0A%0A%23if+(defined(__GNUC__)+%7C%7C+defined(__clang__)+%7C%7C+++++++++++++++++++++++++%5C%0A++defined(__IAR_SYSTEMS_ICC__))+%26%26++++++++++++++++++++++++++++++++++++++%5C%0A++defined(__STDC_VERSION__)+%26%26+(__STDC_VERSION__+%3E%3D+199901L)%0A%23define+TEMP_PRINTF(fmt,+arg)+%5C%0A++__attribute__((format(printf,+fmt,+arg)))%0A%23else%0A%23error%0A%23endif%0A%0Aint+my_msnprintf(char+*buffer,+size_t+maxlength,%0A+++++++++++++++++const+char+*format,+...)%0A++TEMP_PRINTF(3,+4)%3B%0A+++++++++%0Aint+my_msnprintf(char+*buffer,+size_t+maxlength,%0A+++++++++++++++++const+char+*format,+...)%0A%7B%0A++int+retcode%3B%0A++va_list+ap_save%3B+/*+argument+pointer+*/%0A++va_start(ap_save,+format)%3B%0A++retcode+%3D+snprintf(buffer,+maxlength,+format,+ap_save)%3B%0A++va_end(ap_save)%3B%0A++return+retcode%3B%0A%7D%0A%0A+%23ifdef+__GNUC__%0A+%23pragma+GCC+diagnostic+push%0A+%23pragma+GCC+diagnostic+ignored+%22-Wformat%22%0A//%23pragma+GCC+diagnostic+ignored+%22-Wformat-overflow%22%0A+%23pragma+GCC+diagnostic+ignored+%22-Wformat-extra-args%22%0A+%23endif%0A%0Aint+main(void)+%7B%0A++++char+buf%5B100%5D%3B%0A++++my_msnprintf(buf,+sizeof(buf),+%22%25*f%22,+INT_MIN,+9.1)%3B%0A%7D%0A%0A+%23ifdef+__GNUC__%0A+%23pragma+GCC+diagnostic+pop%0A+%23endif%0A'),l:'5',n:'1',o:'C+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cg91,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-Werror-implicit-function-declaration+-O2+-Wno-system-headers+-pthread+-Werror+-pedantic-errors',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+gcc+9.1+(Editor+%231)',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+9.1',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+9.1+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

It could be a `#pragma GCC diagnostic push` without a matching `pop`, but can't
see any, and those only suppress warnings anyway.

It can be a pragma inside system header, but that seems unlikely.

145f87b9e89f3a5e287233fe7d3cf57aca23dd8c changed the default for this warning
from enable to suppress. But the `no-` option remains guarded with the "picky warning" 
condition, meaning it needs `--enable-warnings` or `--enable-debug` to kick in, which
aren't set for this build. (Makes me think this option could be deleted from there.):

It'd be interesting to find out why this warnings gets enabled in the first place, but
the simplest and surest fix would be this (I cannot test it):
```
--- a/tests/libtest/lib557.c
+++ b/tests/libtest/lib557.c
@@ -40,6 +40,7 @@
 #ifdef __GNUC__
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wformat"
+#pragma GCC diagnostic ignored "-Wformat-overflow"
 #pragma GCC diagnostic ignored "-Wformat-extra-args"
 #endif
```

**Comment by vszakats** [comment]
CI confirms the merged patch #15012 working:
https://app.circleci.com/pipelines/github/curl/curl/10243/workflows/d9454560-9ee4-4967-9c69-e7b924823c9f/jobs/107806

It would still be interesting to know what enabled this warning there after bundling tests.

**Comment by vszakats** [comment]
Reproduced this issue in GHA, with a control job using test bundles disabled.

Both reproduced the issue: https://github.com/curl/curl/actions/runs/11002168906

Meaning it wasn't caused by bundling tests into a single binary:

It was however caused by a change to `lib557.c` committed in the test bundle PR #14772:
https://github.com/curl/curl/pull/14772/files#diff-d8400524348193c92a42a1fc8305b933d087f177724b435964da8baf8c59362e

Before that patch, format checks were disabled completely for the whole file.
Since per-file solutions like that don't work with bundled ("unity"-style) builds,
the patch left the format checks enabled and replaced it with per-file warning
suppression. But, all GHA jobs use `--enable-warnings` and `--enable-warnings`
triggered a global suppression for one particular warning: `-Wno-format-overflow`.
It means there was no warning while testing, so I did not explicitly suppress it.

The Circle CI job was the first one to catch this (rare) case.

The fix in #15012 is correct and also fixes `unit1652` which had the same issue.
It also drop the global `-Wno-format-overflow` option, so this won't reoccure.

What's still unknown is why this warning is enabled with real GCC (9.1, 11.0),
but not on godbolt. That's likely just my bothed minimal-repro attempt.
